File: Packages/BeeUrEiCore/Sources/BeeUrEiCore/ChinaCoord.swift
import Foundation

/// WGS-84 ↔ GCJ-02 坐标纠偏（国内导航必需）。
/// iPhone GPS 给的是 WGS-84；高德路线/地点全是 GCJ-02（国家测绘加密偏移）。
/// 不转换则定位点相对路线系统性偏移 100–700 米，逐向引导/偏航检测完全不可用。
/// 算法为公开的标准椭球偏移近似（精度 ~1–2m，远小于 GPS 本身误差）；境外坐标原样返回。
public enum ChinaCoord {
    /// WGS-84 → GCJ-02（把 GPS 定位换算到高德坐标系）。
    public static func wgs84ToGcj02(lat: Double, lon: Double) -> (lat: Double, lon: Double) {
        guard isInChina(lat: lat, lon: lon) else { return (lat, lon) }
        let (dLat, dLon) = delta(lat: lat, lon: lon)
        return (lat + dLat, lon + dLon)
    }

    /// GCJ-02 → WGS-84（一次迭代精化逆变换，误差降到厘米级）。
    public static func gcj02ToWgs84(lat: Double, lon: Double) -> (lat: Double, lon: Double) {
        guard isInChina(lat: lat, lon: lon) else { return (lat, lon) }
        // 初值：反减一次 delta；再用正变换的残差修正一次（标准迭代逆）。
        var wLat = lat, wLon = lon
        for _ in 0..<2 {
            let g = wgs84ToGcj02(lat: wLat, lon: wLon)
            wLat -= g.lat - lat
            wLon -= g.lon - lon
        }
        return (wLat, wLon)
    }

    /// 粗判是否在中国大陆范围（GCJ 偏移仅在境内生效；港澳台/境外不偏移）。
    public static func isInChina(lat: Double, lon: Double) -> Bool {
        // 大陆粗包络；剔除明显境外。边界处 GPS 精度远大于纠偏误差，粗判足够。
        lon >= 72.004 && lon <= 137.8347 && lat >= 0.8293 && lat <= 55.8271
    }

    // MARK: - 标准偏移算法

    private static let a = 6378245.0               // 克拉索夫斯基椭球长半轴
    private static let ee = 0.00669342162296594323 // 偏心率平方

    private static func delta(lat: Double, lon: Double) -> (Double, Double) {
        let dLat0 = transformLat(x: lon - 105.0, y: lat - 35.0)
        let dLon0 = transformLon(x: lon - 105.0, y: lat - 35.0)
        let radLat = lat / 180.0 * .pi
        var magic = sin(radLat)
        magic = 1 - ee * magic * magic
        let sqrtMagic = sqrt(magic)
        let dLat = (dLat0 * 180.0) / ((a * (1 - ee)) / (magic * sqrtMagic) * .pi)
        let dLon = (dLon0 * 180.0) / (a / sqrtMagic * cos(radLat) * .pi)
        return (dLat, dLon)
    }

    private static func transformLat(x: Double, y: Double) -> Double {
        var ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * .pi) + 20.0 * sin(2.0 * x * .pi)) * 2.0 / 3.0
        ret += (20.0 * sin(y * .pi) + 40.0 * sin(y / 3.0 * .pi)) * 2.0 / 3.0
        ret += (160.0 * sin(y / 12.0 * .pi) + 320 * sin(y * .pi / 30.0)) * 2.0 / 3.0
        return ret
    }

    private static func transformLon(x: Double, y: Double) -> Double {
        var ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * .pi) + 20.0 * sin(2.0 * x * .pi)) * 2.0 / 3.0
        ret += (20.0 * sin(x * .pi) + 40.0 * sin(x / 3.0 * .pi)) * 2.0 / 3.0
        ret += (150.0 * sin(x / 12.0 * .pi) + 300.0 * sin(x / 30.0 * .pi)) * 2.0 / 3.0
        return ret
    }
}

File: server/src/nav/chinaCoord.ts
/// WGS-84 ↔ GCJ-02 坐标纠偏（TS 端口，与 iOS BeeUrEiCore/ChinaCoord.swift 同算法）。
/// 用途：高德地理编码(amapGeocode)返回 GCJ-02，而全栈存 WGS-84——保存"家/公司"坐标供围栏判定时须转回 WGS-84，
/// 与客户端上报的 WGS-84 实时位置同系比较。境外坐标原样返回（GCJ 偏移仅大陆生效）。

const A = 6378245.0                // 克拉索夫斯基椭球长半轴
const EE = 0.00669342162296594323 // 偏心率平方

/// 粗判大陆范围（GCJ 偏移仅境内生效；港澳台/境外不偏移）。
export function isInChina(lat: number, lon: number): boolean {
  return lon >= 72.004 && lon <= 137.8347 && lat >= 0.8293 && lat <= 55.8271
}

function transformLat(x: number, y: number): number {
  let ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * Math.sqrt(Math.abs(x))
  ret += ((20.0 * Math.sin(6.0 * x * Math.PI) + 20.0 * Math.sin(2.0 * x * Math.PI)) * 2.0) / 3.0
  ret += ((20.0 * Math.sin(y * Math.PI) + 40.0 * Math.sin((y / 3.0) * Math.PI)) * 2.0) / 3.0
  ret += ((160.0 * Math.sin((y / 12.0) * Math.PI) + 320 * Math.sin((y * Math.PI) / 30.0)) * 2.0) / 3.0
  return ret
}

function transformLon(x: number, y: number): number {
  let ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * Math.sqrt(Math.abs(x))
  ret += ((20.0 * Math.sin(6.0 * x * Math.PI) + 20.0 * Math.sin(2.0 * x * Math.PI)) * 2.0) / 3.0
  ret += ((20.0 * Math.sin(x * Math.PI) + 40.0 * Math.sin((x / 3.0) * Math.PI)) * 2.0) / 3.0
  ret += ((150.0 * Math.sin((x / 12.0) * Math.PI) + 300.0 * Math.sin((x / 30.0) * Math.PI)) * 2.0) / 3.0
  return ret
}

function delta(lat: number, lon: number): { dLat: number; dLon: number } {
  const dLat0 = transformLat(lon - 105.0, lat - 35.0)
  const dLon0 = transformLon(lon - 105.0, lat - 35.0)
  const radLat = (lat / 180.0) * Math.PI
  let magic = Math.sin(radLat)
  magic = 1 - EE * magic * magic
  const sqrtMagic = Math.sqrt(magic)
  const dLat = (dLat0 * 180.0) / (((A * (1 - EE)) / (magic * sqrtMagic)) * Math.PI)
  const dLon = (dLon0 * 180.0) / ((A / sqrtMagic) * Math.cos(radLat) * Math.PI)
  return { dLat, dLon }
}

/// WGS-84 → GCJ-02。
export function wgs84ToGcj02(lat: number, lon: number): { lat: number; lon: number } {
  if (!isInChina(lat, lon)) return { lat, lon }
  const { dLat, dLon } = delta(lat, lon)
  return { lat: lat + dLat, lon: lon + dLon }
}

/// GCJ-02 → WGS-84（两次迭代逆变换，误差降到厘米级）。
export function gcj02ToWgs84(lat: number, lon: number): { lat: number; lon: number } {
  if (!isInChina(lat, lon)) return { lat, lon }
  let wLat = lat
  let wLon = lon
  for (let i = 0; i < 2; i++) {
    const g = wgs84ToGcj02(wLat, wLon)
    wLat -= g.lat - lat
    wLon -= g.lon - lon
  }
  return { lat: wLat, lon: wLon }
}
